File: М10_NodeVMS/nodevms/apphost/reconciler.py
"""Lesson 21 — the reconcile loop, with nothing in it.

Pure logic over an injected store and actuator. No database, no GStreamer,
no network. This is the part of the product that survives the rewrite
(Lesson 24, Step 5): only the actuator changes.

    Desired state is persisted. Actual state is derived.
"""
from __future__ import annotations

import random
from typing import Callable, Protocol

CONVERGED, LAGGING, STALLED = "converged", "lagging", "stalled"


class Store(Protocol):
    def desired(self) -> list[dict]: ...


Actuator = Callable[[str, dict], bool]


class Reconciler:
    def __init__(self, store: Store, actuator: Actuator,
                 max_backoff: float = 60.0, stall_failures: int = 3):
        self.store = store
        self.actuator = actuator
        self.actual: dict[int, dict] = {}     # camera_id -> {"revision": n}   IN MEMORY ONLY
        self.failures: dict[int, dict] = {}   # camera_id -> {"n":, "retry_at":, "delay":}
        self.max_backoff = max_backoff
        self.stall_failures = stall_failures

    def reconcile(self, now: float = 0.0) -> list[tuple[str, int]]:
        desired = {c["id"]: c for c in self.store.desired() if c["enabled"]}
        actions: list[tuple[str, int]] = []

        for cid, cam in desired.items():
            have = self.actual.get(cid)
            if have and have["revision"] >= cam["revision"]:
                continue                                   # already applied
            if self.failures.get(cid) and now < self.failures[cid]["retry_at"]:
                continue                                   # in backoff, not yet
            verb = "start" if not have else "restart"
            if self.actuator(verb, cam):
                self.actual[cid] = {"revision": cam["revision"]}
                self.failures.pop(cid, None)
                actions.append((verb, cid))
            else:
                self._fail(cid, now)
                actions.append(("failed", cid))

        # The stop loop walks what we are RUNNING, not what is desired: you
        # cannot learn about a deletion by looking at rows that exist.
        for cid in list(self.actual):
            if cid not in desired:
                self.actuator("stop", {"id": cid})
                del self.actual[cid]
                actions.append(("stop", cid))
        return actions

    def _fail(self, cid: int, now: float) -> None:
        n = self.failures.get(cid, {}).get("n", 0) + 1
        base = min(2 ** n, self.max_backoff)
        delay = base * (0.5 + random.random() * 0.5)      # jitter: 50-100% of base
        self.failures[cid] = {"n": n, "retry_at": now + delay, "delay": delay}

    def lost(self, cid: int, now: float) -> None:
        """A running pipeline died (bus error, watchdog). Forget it so the next
        pass restarts it, and count the failure so backoff applies."""
        self.actual.pop(cid, None)
        self._fail(cid, now)

    def status(self) -> dict[int, tuple[str, int]]:
        """camera_id -> (position, lag). The vocabulary from Lesson 21, Step 7.
        Positions only — reasons live on the conditions axis (Lesson 24)."""
        out: dict[int, tuple[str, int]] = {}
        for cam in self.store.desired():
            if not cam["enabled"]:
                continue
            cid = cam["id"]
            have = self.actual.get(cid, {}).get("revision", 0)
            lag = max(cam["revision"] - have, 0)
            if lag == 0:
                out[cid] = (CONVERGED, 0)
            elif self.failures.get(cid, {}).get("n", 0) >= self.stall_failures:
                out[cid] = (STALLED, lag)
            else:
                out[cid] = (LAGGING, lag)
        return out

File: М10_NodeVMS/nodevms-go/reconciler/reconciler.go
// Package reconciler is М10 Lesson 2's loop, in Go.
//
// The port that Lesson 5 says is the whole rewrite: the schema, the loop's
// logic, the state machine, the backoff and jitter policy, and the
// desired/actual contract survive unchanged; only the actuator's
// implementation language changes. This file is the loop. Its tests are the
// Python suite's seven, line for line.
//
//	Desired state is persisted. Actual state is derived.
package reconciler

import (
	"math"
	"math/rand"
	"sort"
)

// Position is the vocabulary from Lesson 2, Step 7. Positions only —
// reasons live on the conditions axis (Lesson 5).
type Position string

const (
	Converged Position = "converged"
	Lagging   Position = "lagging"
	Stalled   Position = "stalled"
)

// Camera is one row of desired state: the operator-owned half of М10's
// cameras table, plus the controller-assigned revision.
type Camera struct {
	ID       int64
	Enabled  bool
	Revision int64
}

// Store hands the reconciler the desired state. It is read on every pass;
// the reconciler keeps nothing of it.
type Store interface {
	Desired() []Camera
}

// Actuator is verb -> success. "start", "restart", "stop". In М10 Lesson 3
// it is GStreamer; here and in the tests it is whatever you inject.
type Actuator func(verb string, cam Camera) bool

// Action is what one pass did.
type Action struct {
	Verb string
	ID   int64
}

type failure struct {
	N       int
	RetryAt float64
	Delay   float64
}

// Reconciler compares two sets and acts on the difference. It does not
// process events.
type Reconciler struct {
	store         Store
	actuator      Actuator
	actual        map[int64]int64 // camera id -> applied revision. IN MEMORY ONLY.
	failures      map[int64]failure
	MaxBackoff    float64
	StallFailures int
	rand          *rand.Rand
}

func New(store Store, actuator Actuator) *Reconciler {
	return &Reconciler{
		store:         store,
		actuator:      actuator,
		actual:        map[int64]int64{}, // a fresh process knows nothing and rediscovers everything
		failures:      map[int64]failure{},
		MaxBackoff:    60,
		StallFailures: 3,
		rand:          rand.New(rand.NewSource(rand.Int63())),
	}
}

// Reconcile runs one pass at time `now` (seconds, any monotonic origin —
// tests pass it explicitly so they control the clock).
func (r *Reconciler) Reconcile(now float64) []Action {
	desired := map[int64]Camera{}
	var order []int64
	for _, c := range r.store.Desired() {
		if c.Enabled {
			desired[c.ID] = c
			order = append(order, c.ID)
		}
	}
	sort.Slice(order, func(i, j int) bool { return order[i] < order[j] })

	var actions []Action
	for _, id := range order {
		cam := desired[id]
		have, running := r.actual[id]
		if running && have >= cam.Revision { // already applied: >=, not ==
			continue
		}
		if f, ok := r.failures[id]; ok && now < f.RetryAt { // in backoff, not yet
			continue
		}
		verb := "start"
		if running {
			verb = "restart"
		}
		if r.actuator(verb, cam) {
			r.actual[id] = cam.Revision
			delete(r.failures, id)
			actions = append(actions, Action{verb, id})
		} else {
			r.fail(id, now)
			actions = append(actions, Action{"failed", id})
		}
	}

	// The stop loop walks what we are RUNNING, not what is desired: you
	// cannot learn about a deletion by looking at rows that exist.
	var stops []int64
	for id := range r.actual {
		if _, want := desired[id]; !want {
			stops = append(stops, id)
		}
	}
	sort.Slice(stops, func(i, j int) bool { return stops[i] < stops[j] })
	for _, id := range stops {
		r.actuator("stop", Camera{ID: id})
		delete(r.actual, id)
		actions = append(actions, Action{"stop", id})
	}
	return actions
}

// fail schedules the next attempt: exponential, capped, and JITTERED —
// 50–100 % of the base — so two hundred cameras that failed together do
// not retry together.
func (r *Reconciler) fail(id int64, now float64) {
	n := r.failures[id].N + 1
	base := math.Min(math.Pow(2, float64(n)), r.MaxBackoff)
	delay := base * (0.5 + r.rand.Float64()*0.5)
	r.failures[id] = failure{N: n, RetryAt: now + delay, Delay: delay}
}

// Lost forgets a pipeline that died (bus error, watchdog) so the next pass
// restarts it, and counts the failure so backoff applies.
func (r *Reconciler) Lost(id int64, now float64) {
	delete(r.actual, id)
	r.fail(id, now)
}

// Status is camera id -> (position, lag).
type Status struct {
	Position Position
	Lag      int64
}

func (r *Reconciler) Status() map[int64]Status {
	out := map[int64]Status{}
	for _, cam := range r.store.Desired() {
		if !cam.Enabled {
			continue
		}
		lag := cam.Revision - r.actual[cam.ID]
		if lag < 0 {
			lag = 0
		}
		switch {
		case lag == 0:
			out[cam.ID] = Status{Converged, 0}
		case r.failures[cam.ID].N >= r.StallFailures:
			out[cam.ID] = Status{Stalled, lag}
		default:
			out[cam.ID] = Status{Lagging, lag}
		}
	}
	return out
}

// Failures exposes the backoff table for tests and the console (read-only copy).
func (r *Reconciler) Failures() map[int64]struct {
	N              int
	RetryAt, Delay float64
} {
	out := map[int64]struct {
		N              int
		RetryAt, Delay float64
	}{}
	for id, f := range r.failures {
		out[id] = struct {
			N              int
			RetryAt, Delay float64
		}{f.N, f.RetryAt, f.Delay}
	}
	return out
}

// Actual exposes the in-memory applied revisions (tests only).
func (r *Reconciler) Actual() map[int64]int64 {
	out := map[int64]int64{}
	for k, v := range r.actual {
		out[k] = v
	}
	return out
}

// SetActual is the mistake, kept only so a test can make it on purpose.
// Do not call it from anything that ships.
func (r *Reconciler) SetActual(saved map[int64]int64) {
	r.actual = saved
}
